### PR TITLE
ATOM-15935 Adding draw to pass support to DynamicDrawContext.

### PR DESCRIPTION
**Pull request by VickyAtAZ**
- The DynamicDrawContext can output to different scopes: Scene, RenderPipeline or RasterPass.
- Updated RasterPass so it can handle not only draw calls from Views but also from DynamicDrawContexts.

Signed-off-by: Tao <qingtao@amazon.com>

### CODE REVIEW

File: Gems/Atom/RPI/Code/Include/Atom/RPI.Public/DynamicDraw/DynamicDrawContext.h [hunk 1]
```diff
@@ -177,12 +184,29 @@ namespace AZ
...
             // Reset cached draw data when frame is end (draw data was submitted)
             void FrameEnd();
 
+            void ReInit();
+
             // Get rhi pipeline state which matches current states
             const RHI::PipelineState* GetCurrentPipelineState();
+                        
+            // structure includes DrawItem and stream and index buffer index
+            static const uint32_t InvalidIndex = static_cast<uint32_t>(-1);
```
Comment (rgba16f) [inline Gems/Atom/RPI/Code/Include/Atom/RPI.Public/DynamicDraw/DynamicDrawContext.h:199]: Nit: Can we add a using so it's a unique type?

File: Gems/Atom/RPI/Code/Include/Atom/RPI.Public/DynamicDraw/DynamicDrawContext.h [hunk 2]
```diff
@@ -229,10 +253,21 @@ namespace AZ
             // Draw variations allowed in this DynamicDrawContext
             DrawStateOptions m_drawStateOptions;
 
-            // For generate output attachment layout and filter draw items
-            Scene* m_scene = nullptr;
+            // DrawLListTag used to help setup PipelineState's output
```
Comment (rgba16f) [inline Gems/Atom/RPI/Code/Include/Atom/RPI.Public/DynamicDraw/DynamicDrawContext.h:256]: spelling error extra L DrawLListTag

File: Gems/Atom/RPI/Code/Include/Atom/RPI.Public/DynamicDraw/DynamicDrawContext.h [hunk 3]
```diff
@@ -229,10 +254,21 @@ namespace AZ
...
+            // and also for submitting draw items to views 
             RHI::DrawListTag m_drawListTag;
 
+            // Output scope related
+            enum class OutputScopeType
+            {
+                Scene = 0,
+                RenderPipeline,
+                RasterPass
+            };
```
Comment (antonmic) [inline Gems/Atom/RPI/Code/Include/Atom/RPI.Public/DynamicDraw/DynamicDrawContext.h:267]: Tiny nit: I would add a "Unset" value just for initialization to help with debugging.

File: Gems/Atom/RPI/Code/Source/RPI.Public/Pass/RasterPass.cpp [hunk 4]
```diff
@@ -132,22 +133,60 @@ namespace AZ
...
+                itemCount += drawList.size();
+            }
+            m_combinedDrawList.resize(itemCount);
+            RHI::DrawItemProperties* currentBuffer = m_combinedDrawList.data();
+            for (auto drawList : drawLists)
+            {
+                memcpy(currentBuffer, drawList.data(), drawList.size()*sizeof(RHI::DrawItemProperties));
+                currentBuffer += drawList.size();
+            }
+            SortDrawList(m_combinedDrawList);
```
Comment (antonmic) [inline Gems/Atom/RPI/Code/Source/RPI.Public/Pass/RasterPass.cpp:186]: The view draw list should already be sorted, so we could optimize here using that knowledge. Are there use cases where the pass will have 10,000+ drawitems from the view and other draw lists as well?

**Review by antonmic — APPROVED** [review]

File: Gems/Atom/RPI/Code/Source/RPI.Public/Pass/RasterPass.cpp (hunk 4 above)
Comment (VickyAtAZ) [reply Gems/Atom/RPI/Code/Source/RPI.Public/Pass/RasterPass.cpp:186]: If there is only one draw list from the view, it won't re-sort it. But if there are other draw lists from dynamic draw system, we have to re-sort the combined list. 

I created a jira to find a better solution for this use case. ATOM-15998

File: Gems/AtomLyIntegration/AtomBridge/Code/Source/PerViewportDynamicDrawManager.cpp [hunk 5]
```diff
@@ -105,7 +105,7 @@ namespace AZ::AtomBridge
             {
                 return nullptr;
             }
-            context = RPI::DynamicDrawInterface::Get()->CreateDynamicDrawContext(pipeline);
+            context = RPI::DynamicDrawInterface::Get()->CreateDynamicDrawContext();
             contextFactoryIt->second(context);
```
Comment (rgba16f) [inline Gems/AtomLyIntegration/AtomBridge/Code/Source/PerViewportDynamicDrawManager.cpp:109]: Does this need a SetOutputScope after the new context create?

**Review by rgba16f — APPROVED** [review]

**Comment by VickyAtAZ** [comment]
Opened another pull request to resolve the DCO issue: https://github.com/o3de/o3de/pull/2248

**Pull request closed without merging**